datasetviewer/stack/StackPresenter.py

```diff
@@ -0,0 +1,361 @@
+            PreviewView. Defaults to None and is assigned to the first key in the dataset in the `set_dict` method.
+            Changed during calls to `set_current_face` when a message is received from the PreviewPresenter indicating
+            a different dictionary element has been selected.
+        _stack_idx (dict): Keeps a record of dataset keys and which index they have on the StackView's widget stack.
+            Defaults to None as it set during `_set_dict`.
+
+    Raises:
+        ValueError: If the StackView or the DimensionViewFactory are None.
+    """
+
+    def __init__(self, stack_view, dim_fact):
+
+        if stack_view is None:
+            raise ValueError("Error: Cannot create StackView when View is None.")
+
+        if dim_fact is None:
+            raise ValueError("Error: Cannot create DimensionViewFactory when View is None.")
+
+        self._view = stack_view
+        self._dim_view_factory = dim_fact
+        self._dict = None
+        self._main_presenter = None
+        self._dim_presenters = {}
+        self._current_stack_face = None
+        self._stack_idx = {}
+
+    def register_master(self, master):
+        """
+        Register the MainViewPresenter as the StackPresenter's master, and subscribe the MainViewPresenter to the
+        StackPresenter.
+
+        Args:
+            master (MainViewPresenter): An instance of a MainViewPresenter.
+        """
+
+        assert (isinstance(master, MainViewPresenterInterface))
+
+        self._main_presenter = master
+        master.subscribe_stack_presenter(self)
+
+    def set_dict(self, dict):
+        """
+        Assigns the _dict variable in the StackPresenter, creates the required amount of Dimension widgets, and then
+        places these widgets on the StackView.
+
+        Args:
+            dict (DataSet): An OrderedDict of xarray Datasets.
+        """
+
+        self._dict = dict
+
+        # Clear the DimensionViews that were on the Stack
+        self._clear_stack()
+
+        # Clear the information about the previous presenters/widgets
+        self._dim_presenters = {}
+        self._stack_idx = {}
+
+        for key in dict.keys():
+
+            # Create a new Stack element for every dataset in the dictionary and obtain its index
+            idx = self._view.create_stack_element()
+
+            # Record which index on the Stack this dataset has
+            self._stack_idx[key] = idx
+
+            # Prepare a dictionary for the presenters that this dataset will use
+            self._dim_presenters[key] = {}
+
+            data = dict[key].data
+
+            # Ignore data with a single dimension as they will not need Dimension widgets
+            if len(data.dims) > 1:
+
+                for i in range(len(data.dims)):
+
+                    # Create a dimension widget
+                    w = self._dim_view_factory.create_widgets(data.dims[i], data.shape[i])
```

Comment:
Looks better. Maybe the contents of this for loop could be pulled into another named method for ease of reading.